Remove hardcoded --mquick from crest calls

lbabetto opened the pull request:
This PR removes the hardcoded `--mquick` flag from all `crest` calls. Tests have been updated accordingly (explicitly setting `--mquick` via the `options` parameter).

Dependency finder was also updated, both for Crest and Orca. For Crest, it no longer looks for `xtbiff` for version 3.0 and above, while the Orca finder was updated to simplify the matching between MPI and Orca version.

Closes #33

ppravatto commented on spycci/core/dependency_finder.py:
I'm not entirely convinced about the need to add a `path` argument to `find_orca_version` (or more generally to `find_PROGRAM_version`). The current implementation assumes the program (e.g. `orca`) is available in the system `$PATH`, which works well in most cases. Am I missing a specific use case that this change is meant to support?

If we decide to support this, I’d suggest keeping the `path` argument optional and falling back to the usual lookup when it's not provided.

This because, looking at the rest of the pull request (especially the tests), I noticed several calls like `find_PROGRAM_version(locate_PROGRAM(...))`. This seems a bit strange, since `locate_PROGRAM()` already calls `find_PROGRAM_version()` internally. Nesting them introduces some circular and confusing logic. Using the nested call still relies on `locate_executable()`, which ends up searching the system’s `$PATH` anyway — so it's not clear what additional capability this change introduces.

---

The only scenario where I could see this possibly being useful is if `orca` (or the `PROGRAM`) is not in `$PATH` (or multiple versions are available), and the user provides a custom path via the `ORCADIR` argument during `OrcaInput` construction. However, this would require the inclusion of version checking on construction that, at the moment, is implemented only on default mode. Furthermore, (and this is something to discuss in the future), I am not even sure that the `ORCADIR` option would work since I don't know whether ORCA itself would correctly locate its own internal executables in such a setup.

lbabetto replied:
I added the `program_finder.py` module to replace `dependency_finder.py` in the last commit. New approach is based on `packaging` for version number matching and unifies all program checks in a single workflow.

With the new implementation it is possible to specify how to look for the version number and how to integrate dependencies (`mpirun` for `orca`, and `xtbiff` for `crest<3.0`). This should make it easy to add new programs in the future.

NOTE: the "get_folder" option was removed, the corresponding checks have been integrated directly in the necessary modules (i.e., for `vmd`) as they were simple `.strip()` calls.

ppravatto commented on spycci/wrappers/crest.py:
```diff
@@ -434,18 +434,21 @@ def qcg_grow(
 
+    CREST_VERSION = find_crest_version(CRESTPATH)
+
+    if CREST_VERSION < "3.0":
```
Same thing here, I would avoid checking version with strings

ppravatto commented on spycci/wrappers/crest.py:
```diff
@@ -554,18 +557,21 @@ def qcg_ensemble(
 
+    CREST_VERSION = find_crest_version(CRESTPATH)
+
+    if CREST_VERSION < "3.0":
```
Same thing here, I would avoid checking version with strings

lbabetto replied:
Yes, agreed.

ppravatto commented on tests/functional/test_fukui.py:
I would stick to a defined orca version for testing. Also here avoid version checking with strings.

lbabetto replied:
For the Orca version I agree, what do you think about the slightly different xtb results? Should we also fix the xtb version at this point or is it okay to loosen the thresholds so that they both work? (I suspect that xtb-6.8 would break things again at this point...)

ppravatto reviewed (changes requested):
I left a few comments. Three main points:

- The introduction of the `path` argument in the `dependency_finder` module is not entirely clear to me (I might be missing something important).
- I would avoid comparing versions as plain strings.
- I’d suggest sticking to a single ORCA version for testing, unless there's a specific need to support multiple versions explicitly.

ppravatto commented on spycci/engines/orca.py:
```diff
@@ -1799,8 +1799,8 @@ def cosmors(
-        orca_version = find_orca_version()
-        if int(orca_version.split(".")[0]) < 6:
+        orca_version = find_orca_version(locate_orca())
+        if orca_version < "6.0.0":
```
Why do you say it breaks down with `6.1.0-f.0`?

From a quick test in the console, it seems that using `split` and checking the major version of Orca (the only thing needed to asses the availability of the COSMO-RS function) should work as expected:

```python
>>> mystr = "6.1.0-f.0"
>>> int(mystr.split(".")[0])
6
```
Am I missing a specific case where this fails?

ppravatto replied:
Of course, I get the change in the `dependency_finder` module that solves this issue:

> With the current implementation (even adding 6.1.* support), I get this error:
> 
> ```
>         # Check if the available version meets the requirements
>         openmpi_required = {"6.1.*": ["4.1.6"], "6.0.*": ["4.1.6"], "5.0.*": ["4.1.1"], "4.2.*": ["3.1.4"], "4.1.*": ["3.1.3", "2.1.5"]}
> 
>         key = ".".join(orca_version.split(".")[0:-1] + ["*"])
> >       if openmpi_version not in openmpi_required[key]:
>                                   ^^^^^^^^^^^^^^^^^^^^^
> E       KeyError: '6.1.0-f.*'
> 
> spycci/core/dependency_finder.py:123: KeyError
> ```

But I still don't undestand why the change in the COSMO-RS function is needed. If we are just comparing the major release why should we substitute: `if int(orca_version.split(".")[0]) < 6:` with `if orca_version < "6.0.0":`?

lbabetto commented:
I replaced the old `dependency_finder.py` with the new one, we can keep that name for now since it's just for internal use. I also added a unit test suite that should cover most of the cases, using a monkeypatch approach to simulate the expected output (so they should work even if the programs are not actually installed).

At this point "real" integration tests should be redundant, no? Let me know if I should still add some anyway.

ppravatto reviewed (approved):
Looks good, nice job!
